### DISCUSSION TITLE
Webhook with POST and GET methods at the sametime

### DISCUSSION THREAD

**Discussion by pedrojunqueira** [fastapi/fastapi#7001] [Questions] [answered]
Hi I created a route in Flask app to handle webhook which is accept both POST and GET request and will handle it depending on the request method sent to the route.

here is the "happy path" of a Flask route 

```python
@app.route('/webhook', methods=['GET','POST'])
def webhook():
    if request.method == 'POST':
        # request.json #dosomething
        return 'EVENT_RECEIVED' , 200
    else:

        response = dict()

        response["hub.challenge"] = request.args.get('hub.challenge')

        return response, 200

```

I am trying the same with FAST API using the Request starlette class and the Basemodel I wanted to make both optional and handle separately. Like checking the Request object first if not then get the request body of a post request.

```python
@app.post("/webhook")
@app.get("/webhook")
async def webhook(request: Request, event: Event):

    challenge = request.query_params.get('hub.challenge')

    if challenge:
        return {"hub.challenge": challenge}


    event #dosomething
    return 'EVENT_RECEIVED' , 200

```

Here is a get request example

```python
import requests

url = "http://127.0.0.1:8000/webhook?hub.verify_token=STRAVA&hub.challenge=15f7d1a91c1f40f8a748fd134752feb3&hub.mode=subscribe"

r = requests.get(url)


```

Here is a post request example

```python
import requests

event = {
    "aspect_type": "update",
    "event_time": 1516126040,
    "object_id": 1360128428,
    "object_type": "activity",
    "owner_id": 134815,
    "subscription_id": 120475,
    "updates": {
        "title": "Messy"
    }
}

r = requests.post("http://127.0.0.1:8000/webhook", json=event)
```

The post is working but when I send a get request here is the IO response

```bash
{'date': 'Mon, 01 Feb 2021 11:29:28 GMT', 'server': 'uvicorn', 'content-length': '81', 'content-type': 'application/json'}
{"detail":[{"loc":["body"],"msg":"field required","type":"value_error.missing"}]}
422
```

FastAPI Logs
```bash
INFO:     127.0.0.1:56956 - "GET /webhook?hub.verify_token=STRAVA&hub.challenge=15f7d1a91c1f40f8a748fd134752feb3&hub.mode=subscribe HTTP/1.1" 422 Unprocessable Entity
```

Thanks for assistance on this

:)

### First check

* [ X] I added a very descriptive title to this issue.
* [ X] I used the GitHub search to find a similar issue and didn't find it.
* [ X] I searched the FastAPI documentation, with the integrated search.
* [ X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ X] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

**Comment by Mause** [comment]
Have you looked at the documentation regarding optional fields/query params?

**Comment by insomnes** [comment]
@pedrojunqueira you can use `app.api_route` decorator for multiple methods and also should set event as Optional:
```
@app.api_route("/webhook", methods=["GET", "POST"])
async def webhook(request: Request, event: Optional[Event] = None):

    if request.method == "GET":
        return "EVENT_RECEIVED_GET"

    elif request.method == "POST":
        return "EVENT_RECEIVED_POST"
```
This will not demand JSON with POST request also, you should check it yourself in functiuon.
But i suggest you to change you decision to separated GET and POST functions it would be much easier to maintain and read for some complicated logic.

Also here `return 'EVENT_RECEIVED' , 200` will not set the status code of the response, you will just return response as JSON with `["EVENT_RECEIVED", 200]`. For response status code read docs here:
https://fastapi.tiangolo.com/tutorial/response-status-code/
and here
https://fastapi.tiangolo.com/advanced/additional-status-codes/

And as @Mause mentioned, for query parameters read the docs:
https://fastapi.tiangolo.com/tutorial/query-params/

**Comment by pedrojunqueira** [comment, marked as answer]
Thanks @insomnes  

final code worked fine. See below

```python
from typing import Optional


from fastapi import FastAPI, Request
from pydantic import BaseModel

app = FastAPI()


class Event(BaseModel):
    aspect_type: str
    event_time: int
    object_id: int
    object_type: str
    owner_id: int
    subscription_id: int
    updates: dict

@app.api_route("/webhook",methods=["GET", "POST"])
async def read_params(request: Request, event: Optional[Event] = None):

    if request.method == 'GET':

        challenge = request.query_params.get('hub.challenge')

        print(challenge) # do something with get request data

        if challenge:
            return {"hub.challenge": request.query_params.get('hub.challenge')}

    print(event) # do something with event data

    return 'EVENT_RECEIVED' 
```

I should have investigated the Request class better. But in the end your answer gave a better understanding on FAST_API

I am going to close the issue.

Cheers

**Comment by tiangolo** [comment]
Thanks for the help here @insomnes and @Mause ! 👏  🙇 

Thanks for reporting back and closing the issue @pedrojunqueira  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
